### PR TITLE
Select module op

### PR DESCRIPTION
**Pull request by CharlieL7**
### Main Info
* Creates the `select_module` operator that selects one of the submodules passed to it to run based on the submodule parameters
  * The submodule is selected by having the exact same static shapes for the arguments to `select_module` as the parameters in the submodule
* This op is planned to be used with dynamic batching and I made dynamic batch test cases for test_verify and ref_ops_test
* Added a new function to gpu::lowering, `add_select_module_op()` that sets the new module attribute `use_local_alloc` to true in the submodules and adds a `hip::sync_stream` before the submodule return
  * The reasoning for this is to avoid using output parameters between the main module and the submodules
  * Change to `replace_allocate` pass to not add output parameters if `module.use_local_alloc` is true
  * Trying to pass output parameters from the main module leads to an allocation of the maximum memory size and multiple output parameters. It makes more sense to have the submodules do the allocation
  * It is not clear to me why `hip::sync_stream` gets around the `dangling reference` issue. What I coded seems to work, but I do not understand when the copy to host occurs.

### Other info
* ~Why was `context.hpp` changed to use `std::move(queue)`? This is  the result of `make generate`~
* Changes to `run_verify` to handle dynamic shape parameters. Creates a static shape with the maximum dimensions.
* New function to shape `shape.any_of_dynamic()` to handle checking if subshapes are dynamic

### Sample compile output from `test_select_module_reduce` verify test
```
module: "main"
main:@0 = check_context::migraphx::version_1::gpu::context -> float_type, {}, {}
main:@1 = hip::hip_allocate_memory[shape=float_type, {0}, {1},id=main:scratch] -> float_type, {0}, {1}
data = @param:data -> dynamic, float_type, {[1, 4, 0], [2, 2, 0], [2, 2, 0]}
main:@3 = select_module[output_dyn_shapes=[dynamic, float_type, {[1, 4, 0], [2, 2, 0]}]](data), [batch_1, batch_2, batch_3, batch_4] -> [dynamic, float_type, {[1, 4, 0], [2, 2, 0]}]
main:@4 = get_tuple_elem[index=0](main:@3) -> dynamic, float_type, {[1, 4, 0], [2, 2, 0]}
main:@5 = @return(main:@4)

module: "batch_1"
batch_1:@0 = check_context::migraphx::version_1::gpu::context -> float_type, {}, {}
batch_1:@1 = hip::hip_allocate_memory[shape=float_type, {2}, {1},id=batch_1:scratch] -> float_type, {2}, {1}
batch_1:@2 = load[offset=0,end=8](batch_1:@1) -> float_type, {1, 1, 2}, {2, 2, 1}
data = @param:data -> float_type, {1, 2, 2}, {4, 2, 1}
batch_1:@4 = gpu::code_object[code_object=13632,symbol_name=reduce_kernel,global=128,local=64,](data,batch_1:@2) -> float_type, {1, 1, 2}, {2, 2, 1}
batch_1:@5 = squeeze[axes={1}](batch_1:@4) -> float_type, {1, 2}, {2, 1}
batch_1:@6 = hip::sync_stream(batch_1:@5) -> float_type, {1, 2}, {2, 1}
batch_1:@7 = @return(batch_1:@6)

module: "batch_2"
batch_2:@0 = check_context::migraphx::version_1::gpu::context -> float_type, {}, {}
batch_2:@1 = hip::hip_allocate_memory[shape=float_type, {4}, {1},id=batch_2:scratch] -> float_type, {4}, {1}
batch_2:@2 = load[offset=0,end=16](batch_2:@1) -> float_type, {2, 1, 2}, {2, 2, 1}
data = @param:data -> float_type, {2, 2, 2}, {4, 2, 1}
batch_2:@4 = gpu::code_object[code_object=13632,symbol_name=reduce_kernel,global=256,local=64,](data,batch_2:@2) -> float_type, {2, 1, 2}, {2, 2, 1}
batch_2:@5 = squeeze[axes={1}](batch_2:@4) -> float_type, {2, 2}, {2, 1}
batch_2:@6 = hip::sync_stream(batch_2:@5) -> float_type, {2, 2}, {2, 1}
batch_2:@7 = @return(batch_2:@6)

module: "batch_3"
batch_3:@0 = check_context::migraphx::version_1::gpu::context -> float_type, {}, {}
batch_3:@1 = hip::hip_allocate_memory[shape=float_type, {6}, {1},id=batch_3:scratch] -> float_type, {6}, {1}
batch_3:@2 = load[offset=0,end=24](batch_3:@1) -> float_type, {3, 1, 2}, {2, 2, 1}
data = @param:data -> float_type, {3, 2, 2}, {4, 2, 1}
batch_3:@4 = gpu::code_object[code_object=13632,symbol_name=reduce_kernel,global=384,local=64,](data,batch_3:@2) -> float_type, {3, 1, 2}, {2, 2, 1}
batch_3:@5 = squeeze[axes={1}](batch_3:@4) -> float_type, {3, 2}, {2, 1}
batch_3:@6 = hip::sync_stream(batch_3:@5) -> float_type, {3, 2}, {2, 1}
batch_3:@7 = @return(batch_3:@6)

module: "batch_4"
batch_4:@0 = check_context::migraphx::version_1::gpu::context -> float_type, {}, {}
batch_4:@1 = hip::hip_allocate_memory[shape=float_type, {8}, {1},id=batch_4:scratch] -> float_type, {8}, {1}
batch_4:@2 = load[offset=0,end=32](batch_4:@1) -> float_type, {4, 1, 2}, {2, 2, 1}
data = @param:data -> float_type, {4, 2, 2}, {4, 2, 1}
batch_4:@4 = gpu::code_object[code_object=13632,symbol_name=reduce_kernel,global=512,local=64,](data,batch_4:@2) -> float_type, {4, 1, 2}, {2, 2, 1}
batch_4:@5 = squeeze[axes={1}](batch_4:@4) -> float_type, {4, 2}, {2, 1}
batch_4:@6 = hip::sync_stream(batch_4:@5) -> float_type, {4, 2}, {2, 1}
batch_4:@7 = @return(batch_4:@6)
```

### CODE REVIEW

**Comment by umangyadav** [comment]
> Why was context.hpp changed to use std::move(queue)? This is the result of make generate
 If you rebase and do `make generate` it would go away now.

File: src/include/migraphx/op/select_module.hpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                assert(param_names.size() <= args.size());
+                return std::equal(param_names.cbegin(),
+                                  param_names.cend(),
+                                  args.cbegin(),
+                                  [&](auto p_name, auto a) {
+                                      return a.get_shape() == mr->get_parameter_shape(p_name);
+                                  });
+            });
+
+        if(module_iter == submodule_list.end())
```
Comment (TedThemistokleous) [inline src/include/migraphx/op/select_module.hpp:71]: Add a test case to cover this

File: src/replace_allocate.cpp [hunk 2]
```diff
@@ -104,19 +104,17 @@ void replace_allocate::apply(module& m) const
             continue;
 
         auto s = ins->get_shape();
-
-        if(not main_offload_copy and model.needs_out_params() and contains(mod_output_names, ins))
+        if(not main_offload_copy and not(m.use_local_alloc) and model.needs_out_params() and
```
Comment (TedThemistokleous) [inline src/replace_allocate.cpp:107]: Just take the compliment of the set of either outcome for allocs

not main_offload_copy and not(m.use_local_alloc) - > not(main_offload_copy or m.use_local_alloc)

**Review by TedThemistokleous — CHANGES_REQUESTED** [review]
Some comments /questions. 

Resolve the initial case as to why we;re seeing the zero alloc in your example earlier.

```
module: "main"
main:@0 = check_context::migraphx::version_1::gpu::context -> float_type, {}, {}
**main:@1 = hip::hip_allocate_memory[shape=float_type, {0}, {1},id=main:scratch] -> float_type, {0}, {1}**
data = @param:data -> dynamic, float_type, {[1, 4, 0], [2, 2, 0], [2, 2, 0]}
main:@3 = select_module[output_dyn_shapes=[dynamic, float_type, {[1, 4, 0], [2, 2, 0]}]](data), [batch_1, batch_2, batch_3, batch_4] -> [dynamic, float_type, {[1, 4, 0], [2, 2, 0]}]
main:@4 = get_tuple_elem[index=0](main:@3) -> dynamic, float_type, {[1, 4, 0], [2, 2, 0]}
main:@5 = @return(main:@4)

```

Seems benign, but I sense this is a placeholder for the other scratch pads for memory for each module.

**Review by kahmed10 — APPROVED** [review]

**Review by TedThemistokleous — APPROVED** [review]

**Review by umangyadav — APPROVED** [review]

**Comment by pfultz2** [comment]
> It is not clear to me why hip::sync_stream gets around the dangling reference issue.

`hip::sync_stream` is missing an `output_alias`, which is why it skips the checks.

> What I coded seems to work, but I do not understand when the copy to host occurs.

It might dangle when the user destroys the program and wants to still use the output buffer. Even if we shared the ownership to avoid the dangling reference, this would still require the scratch memory to stay alive just to use the output buffer, which we shouldnt be doing.

This should be using outputs parameters and allocations like we do for `if` and `loop` operators.

**Pull request merged**